File: 프로그래머스/lv1/42889. 실패율/실패율.py
def solution(N, stages):
    map = {}
    answer = []
    set_stages = set(stages)
    
    for i in range(1, N+2):
        map[i] = [0, 0]

    for s in set_stages:
        count = stages.count(s)
        for i in range(1, s + 1):
            if i == s:
                map[i][0] += count
            map[i][1] += count
    
    for n in range(1, N + 1):
        trial = map[n][1]
        failure = map[n][0]
        if trial == 0:
            answer.append((0, n))
        else:
            answer.append((failure/trial, n))

    answer.sort(key = lambda x: (-x[0], x[1]))
    return [x[1] for x in answer]

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    var answer = [];
    const stageClear = new Map()
    const stageFail = new Map()
    
    for (let i = 1; i <= N; i++) {
        stageClear.set(i, 0)
        stageFail.set(i, 0)
    }
    
    for (let i = 0; i < stages.length; i++) {
        const stage = stages[i]
        stageFail.set(stage, stageFail.get(stage) + 1)
        for (let j = 1; j <= stage - 1; j++) {
            stageClear.set(j, stageClear.get(j) + 1)
        }
    }
    
    for (let i = 1; i <= N; i++) {
        const rate = stageFail.get(i) / stageClear.get(i)
        answer.push([i, rate])
    }
    
    answer = answer.sort((a, b) => a[1] === b[1] ? a[0] - b[0] : b[1] - a[1]).map(el => el[0])
    return answer
}